lib/linked_list.rb

```diff
@@ -19,99 +19,228 @@ def initialize
     def reverse
-      raise NotImplementedError
+      return if @head.nil? || @head.next.nil?
+
+      curr = @head
+      prev = nil
+
+      until curr.nil?
+        temp = curr.next
+        curr.next = prev
+        prev = curr
+        curr = temp
+      end
     end
 
 
     ## Advanced Exercises
     # returns the value at the middle element in the singly linked list
     def find_middle_value
-      raise NotImplementedError
+      return nil if @head.nil?
+
+      slow = @head
+      fast = @head.next
+
+      until fast.nil?
+        slow = slow.next
+        fast = fast.next.next
+      end
+
+      return slow.data
     end
 
     # find the nth node from the end and return its value
     # assume indexing starts at 0 while counting to n
     def find_nth_from_end(n)
-      raise NotImplementedError
+      return nil if @head.nil? || self.length < n
+      
+      slow = @head
+      fast = @head
+
+      (n - 1).times do
+        fast = fast.next
+      end 
+
+      until fast.next.nil?
+        slow = slow.next
+        fast = fast.next
+      end
+
+      return slow.data
     end
 
     # checks if the linked list has a cycle. A cycle exists if any node in the
     # linked list links to a node already visited.
     # returns true if a cycle is found, false otherwise.
     def has_cycle
-      raise NotImplementedError
+      return false if @head.nil?
+      marker = @head.next
+
+      until marker.nil? 
+        return true if marker == @head
+        marker = marker.next
+      end
+      return false
     end
 
 
     # Additional Exercises 
     # returns the value in the first node
     # returns nil if the list is empty
     def get_first
-      raise NotImplementedError
+      return nil if @head.nil?
+      return @head.data
     end
 
     # method that inserts a given value as a new last node in the linked list
     def add_last(value)
-      raise NotImplementedError
+      if @head.nil?
+        @head = Node.new(value)
+      end
+
+      curr = @head
+      self.(length-1).times do
+        curr = curr.next
+      end
+      curr.next = Node.new(value)
     end
 
     # method that returns the value of the last node in the linked list
     # returns nil if the linked list is empty
     def get_last
-      raise NotImplementedError
+      return nil if @head.nil?
+      
+      curr = @head
+      until curr.next.nil?
+        curr = curr.next
+      end
+
+      return curr.data
     end
 
     # method to insert a new node with specific data value, assuming the linked
     # list is sorted in ascending order
     def insert_ascending(value)
-      raise NotImplementedError
+      @head = Node.new(value) if @head.nil?
+
+      if @head.data > value
+        temp = @head
+        @head = Node.new(value)
+        @head.next = temp
+      else
+        curr = @head 
+        until curr.nil?
+          if value >= curr.data
+            new = Node.new(value,curr.next.next)
+            curr.next = new

```

Comment:
You'll want to leave the loop when you add the new node. 

```suggestion
            curr.next = new
            return
```